lib/loopdev.c

```diff
@@ -1623,18 +1624,23 @@ int loopcxt_ioctl_blocksize(struct loopdev_cxt *lc, uint64_t blocksize)
 int loopcxt_delete_device(struct loopdev_cxt *lc)
 {
 	int rc, fd = loopcxt_get_fd(lc);
+	int lock_fd = -1;
 
 	if (fd < 0)
 		return -EINVAL;
 
+	lock_fd = open(lc->device, O_RDONLY|O_CLOEXEC);
```

Comment:
Why do we need extra FD for the device lock? Would it be enough to use `fd` as returned by loopcxt_get_fd(lc)?

Reply:
From one script:

1. We are creating a loop device say loop0. On this loop0, we are creating partitions
2. Then we are formatting the loop0p1 with xfs
3. Reading the blkid of the loop0p1

From another script:

1. Deleting the loop0 device 

Earlier I thought that its only "systemd-udevd" which is creating the issue, but later I realized that its not only the
"systemd-udevd" but also the filesystem utilities like "blkid" is also involved.

Issue with using existing fd for flock()
----------------------------------------
1. When we initiate the loop device deletion using "losetup -d" from user-space, the reference count of the loop device
   (lo->lo_refcnt) will be 1 in kernel, when there is no other process opened the device
2. This will cause the kernel code to proceed ahead with the deletion process:

	static int loop_clr_fd(struct loop_device *lo)
	{
		...
		...
		if (atomic_read(&lo->lo_refcnt) > 1) {
			lo->lo_flags |= LO_FLAGS_AUTOCLEAR;
			mutex_unlock(&loop_ctl_mutex);
			return 0;
		}
		...
		...
	}

3. As part of the loop deletion, at the end, the partitions will be scanned for removing them
4. While dropping these partitions, it will check the opened partition count of the block device (bdev->bd_part_count)
5. This bd_part_count will be incremented if any other process opens the partiton
6. If, before checking the partitions from "losetup -d", "blkid/systemd-udevd" enters and opens the loop0p1, then the
   bd_part_count for loop0 will be incremented
7. Now, when the kernel code proceeds ahead with "drop_partitions()", it will check if any partition is opened or not
8. If any partition is opened, then the kernel code will return EBUSY from drop_partitions() function
9. So, the partition will not be deleted but the whole disk (loop0) will be deleted causing stale partitions remaining

With separate fd for flock()
----------------------------
1. The reference count of the loop device loop0 will be 2 as we have opened the device twice (unlike in point "1" above)
2. Now due to the kernel code shown in point 2 above, the deletion will not proceed ahead
3. Instead, it will set the LO_FLAGS_AUTOCLEAR flag asking to delete when reference count drops to 0
4. So, scanning partitions and the deletion is actually skipped due to reference count > 1
5. If "blkid/systemd-udevd" is entered and opens the loop0p1, the bd_part_count will be incremented for loop0
6. When all processes closes the loop0 and partitions, lo_release() function in kernel will check
   LO_FLAGS_AUTOCLEAR flag when reference count of loop0 drops to 0
7. As this flag is set, it will invoke the loop device deletion (__loop_clr_fd())
8. So, the loop device and all its partitions will be deleted without any issue

So, this issue occurs if
------------------------
1. If the "losetup -d" ran first before "blkid/systemd-udevd"
2. And "blkid/systemd-udevd" gets a chance to run before the kernel scans the partitions of loop0 (for "losetup -d")
3. And before "blkid/systemd-udevd" closes the partition, if "losetup -d" gets a chance to proceed from where it left

This is the reason why we are using the separate fd.

Reply:
Hi Karel,
Can you please review my above comments?